server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -145,8 +136,11 @@ public void run() throws IOException {
                 );
             }
         }
-
-        runCoordinatorRewritePhase();
+        if (getNumShards() == 0) {
+            finishPhase();
```

Comment:
I believe you added this `if` because you removed the same check that happens before at the beginning of the `start` method. As far as I can see, `runCoordinatorRewritePhase` loops over the iterators which does nothing if they are empty, and then calls `finishPhase`. That makes me wonder if we need this.

Reply:
Yes, I removed it in https://github.com/elastic/elasticsearch/pull/95064/commits/e412458102f57ccd3915e02055e7d32a9c2c22c2.